core/domain/html_validation_service.py

```diff
@@ -0,0 +1,890 @@
+                    if arg_type == objects.Filepath:
+                        if not re.match(filename_re, arg_value):
+                            err_msg = "Invalid filepath"
+
+                # To ensure video id has 11 chars.
+                elif tag_name == 'oppia-noninteractive-video':
+                    if arg_type == objects.UnicodeString:
+                        if len(arg_value) != 11:
+                            err_msg = "Video id length is not 11"
+
+                # To ensure that nested collapsible or tab
+                # is not present in collapsible component.
+                elif tag_name == 'oppia-noninteractive-collapsible':
+                    inner_soup = bs4.BeautifulSoup(
+                        arg_value.encode(encoding='utf-8'),
+                        'html.parser')
+                    collapsible_count = len(inner_soup.findAll(
+                        name='oppia-noninteractive-collapsible'))
+                    tabs_count = len(inner_soup.findAll(
+                        name='oppia-noninteractive-tabs'))
+                    if collapsible_count or tabs_count:
+                        err_msg = "Nested tabs and collapsible"
+
+            except Exception as e:
+                err_msg = str(e)
+        else:
+            # This block is to check args for tab component.
+            # Ensure that arg is a list and each element item in
+            # the list is a dictionary.
+            if isinstance(arg_value, list) and (
+                    all(isinstance(item, dict) for item in (
+                        arg_value))):
+                for value in arg_value:
+                    value_keys = value.keys()
+                    # Each item has exactly two keys namely content
+                    # and title.
+                    if set(value_keys) != set(['content', 'title']):
+                        err_msg = "Invalid attributes in tab contents"
+                    else:
+                        content = value['content']
+                        title = value['title']
+                        title_dict = schema['items']['properties'][0]
+                        content_dict = schema['items']['properties'][1]
+                        title_type = obj_types_to_obj_classes[
+                            title_dict['schema']['type']]
+                        content_type = obj_types_to_obj_classes[
+                            content_dict['schema']['type']]
+                        try:
+                            title_type.normalize(title)
+                            content_type.normalize(content)
+                            # To ensure that nested collapsible or tab
+                            # is not present in tabs component.
+                            inner_soup = bs4.BeautifulSoup(
+                                content.encode(encoding='utf-8'),
+                                'html.parser')
+                            collapsible = inner_soup.findAll(
```

Comment:
Don't repeat yourself. Loop over a list of "complex components". I think there's one defined above already.

Reply:
Done!